correctly make RenderNode not a glib::Object

bvinc opened the pull request:
I took a stab at fixing #48.

I made a macro to help the boiler plate of defining a new render node.  But I wanted each render node in its own file, so I ended up exporting the macro, which is unfortunate.  Do you have any other suggestions?  I'm open to getting rid of the macro.

Also, there must be something that I don't understand, because when I run generator.py I get dozens of unrelated files changed.  I hand picked the files that I changed in this PR.

bilelmoussaoui commented on gsk4/src/render_node.rs:
```diff
@@ -0,0 +1,171 @@
+
+        impl ::std::fmt::Display for $rust_type {
+            fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
+                f.write_str($display_name)
```
why not just stringify the rust_type?

bvinc replied:
good idea!

bilelmoussaoui commented on gsk4/Gir.toml:
```diff
@@ -72,86 +70,27 @@ manual = [
-    name = "new"
-    # uses &[&Shadow] instead of a &[Shadow]
+    name = "render"
+    # gir 
```
please make this more descriptive of the reason why it's manual, same for `render_texture`

bvinc replied:
done

bilelmoussaoui commented on gsk4/src/render_node.rs:
```diff
@@ -0,0 +1,171 @@
+
+    pub fn downcast<T: IsARenderNode>(self) -> Result<T, RenderNode> {
+        unsafe {
+            if self.get_type() == T::static_type() {
```
You should probably use get_node_type here, see gdk::Event for an example

bilelmoussaoui commented on gsk4/src/render_node.rs:
```diff
@@ -0,0 +1,171 @@
+        }
+    }
+
+    pub fn get_type(&self) -> glib::Type {
+        Self::static_type()
+    }
```
This doesn't make much sense, as you should be using something else for downcast/downcast_ref

bvinc replied:
Yup, I got rid of that method.

bilelmoussaoui commented on gsk4/src/render_node.rs:
```diff
@@ -0,0 +1,171 @@
+            }
+        }
+
+        unsafe impl Send for $rust_type {}
+        unsafe impl Sync for $rust_type {}
```
Are the render nodes actually Send?

bvinc replied:
I guess I'm not sure.  I removed it so it should be the same as it was before my change.

bilelmoussaoui commented:
Thanks for looking at this one!

bilelmoussaoui commented on gsk4/src/render_node.rs:
```diff
@@ -0,0 +1,171 @@
+                get_type => || $get_type(),
+            }
+        }
+
```
impl StaticType is missing here

bvinc replied:
This is implemented for me by specifying the get_type function.

bvinc commented:
I used an associated constant for each RenderNode type, that's not going to cause a problem like with a minimum rust version, is it?

bilelmoussaoui commented:
Looks good to me, what do you think @sdroege ?

bilelmoussaoui commented:
> I used an associated constant for each RenderNode type, that's not going to cause a problem like with a minimum rust version, is it?

That should be fine as we require 1.48 anyway

sdroege commented:
Also we use associated constants everywhere else already

sdroege commented on gsk4/src/render_node.rs:
```diff
@@ -0,0 +1,161 @@
+impl RenderNode {
+    //#[doc(alias = "gsk_render_node_deserialize")]
+    //pub fn deserialize(bytes: &glib::Bytes, error_func: /*Unimplemented*/FnMut(/*Ignored*/ParseLocation, /*Ignored*/ParseLocation, &glib::Error), user_data: /*Unimplemented*/Option<Fundamental: Pointer>) -> Option<RenderNode> {
+    //    unsafe { TODO: call ffi:gsk_render_node_deserialize() }
```
This looks relatively easy to add, maybe create an issue if you don't want to do it as part of this PR.

bilelmoussaoui replied:
It's handled by #157 already

sdroege commented on gsk4/src/render_node.rs:
```diff
@@ -0,0 +1,161 @@
+        unsafe { from_glib_full(ffi::gsk_render_node_serialize(self.to_glib_none().0)) }
+    }
+
+    pub fn write_to_file(&self, filename: &str) -> Result<(), glib::Error> {
```
Should the filename be UTF-8 or should it maybe be a `std::path::Path`? (Also missing annotation in the C code if that's the case)

bvinc replied:
This was copied from the generated code.  I'm not sure of the requirements of `gsk_render_node_write_to_file` but std::path::Path is probably preferred.  I'll change it.

sdroege commented on gsk4/src/render_node.rs:
```diff
@@ -0,0 +1,161 @@
+        unsafe { from_glib(ffi::gsk_render_node_get_node_type(self.to_glib_none().0)) }
+    }
+
+    pub fn serialize(&self) -> Option<glib::Bytes> {
```
I don't think this can return `None`

bvinc replied:
Also copied from generated code.  I'll change it.

sdroege commented on gsk4/src/render_node.rs:
```diff
@@ -0,0 +1,161 @@
+    }
+}
+
+pub trait IsARenderNode:
```
Not a fan of the name. I would've called it `RenderNode` but that exists already. `RenderNodeType`? Not sure :)

bvinc replied:
To be honest, I'm not a fan of the name either.  Not only does `RenderNode` exist, but also `RenderNodeType` also exists.  At first, I was using the name `RenderNodeSubType`.  Do you have any other suggestions?

bilelmoussaoui replied:
Maybe just `IsRenderNode` ?

sdroege replied:
I mean, `IsRenderNode` sounds better than `IsARenderNode` to me :)

sdroege commented on gsk4/src/render_node.rs:
```diff
@@ -0,0 +1,161 @@
+
+            fn upcast(self) -> crate::RenderNode {
+                unsafe {
+                    from_glib_full(self.to_glib_full() as *mut ffi::GskRenderNode)
```
You can also just transmute here

sdroege commented on gsk4/src/shadow_node.rs:
```diff
@@ -23,4 +29,14 @@ impl ShadowNode {
     }
+
+    #[doc(alias = "gsk_shadow_node_get_child")]
+    pub fn get_child(&self) -> Option<RenderNode> {
```
Can this actually return `None`?

bvinc replied:
fixed

sdroege commented on gsk4/src/radial_gradient_node.rs:
```diff
@@ -29,4 +37,53 @@ impl RadialGradientNode {
     }
+
+    #[doc(alias = "gsk_radial_gradient_node_get_center")]
+    pub fn get_center(&self) -> Option<graphene::Point> {
```
I don't think this can return `None`

sdroege commented on gsk4/src/linear_gradient_node.rs:
```diff
@@ -24,4 +32,38 @@ impl LinearGradientNode {
+    }
+
+    #[doc(alias = "gsk_linear_gradient_node_get_start")]
+    pub fn get_start(&self) -> Option<graphene::Point> {
```
I don't think this can return `None`

bvinc replied:
fixed

sdroege commented on gsk4/src/linear_gradient_node.rs:
```diff
@@ -24,4 +32,38 @@ impl LinearGradientNode {
+    }
+
+    #[doc(alias = "gsk_linear_gradient_node_get_end")]
+    pub fn get_end(&self) -> Option<graphene::Point> {
```
I don't think this can return `None`

sdroege commented on gsk4/src/conic_gradient_node.rs:
```diff
@@ -23,4 +31,38 @@ impl ConicGradientNode {
     }
+
+    #[doc(alias = "gsk_conic_gradient_node_get_center")]
+    pub fn get_center(&self) -> Option<graphene::Point> {
```
I don't think this can return `None`

sdroege commented on gsk4/src/border_node.rs:
```diff
@@ -1,19 +1,35 @@
+    }
+
+    #[doc(alias = "gsk_border_node_get_outline")]
+    pub fn get_outline(&self) -> Option<RoundedRect> {
```
I don't think this can return `None`

sdroege commented on gsk4/src/border_node.rs:
```diff
@@ -1,19 +1,35 @@
+    //}
+
+    #[doc(alias = "gsk_border_node_get_colors")]
+    pub fn get_colors(&self) -> Option<gdk::RGBA> {
```
I don't think this can return `None`

sdroege commented on gsk4/src/auto/renderer.rs:
```diff
@@ -95,30 +84,6 @@ impl<O: IsA<Renderer>> RendererExt for O {
         }
     }
 
-    fn render<P: IsA<RenderNode>>(&self, root: &P, region: Option<&cairo::Region>) {
```
These should probably not all disappear, otherwise the whole of GSK becomes useless.

Can you add a manual implementation for them into `src/renderer.rs` into a `RendererExtManual` trait?

bilelmoussaoui replied:
Ah good catch. I assumed it was already the case from the `Gir.toml` as it mentions `RendererManualExt` (which is wrong btw)

sdroege reviewed (commented):
Looks mostly ok to me, thanks!

bvinc commented:
I addressed all of the feedback, except using transmute on the `upcast` method.  I'm not confident enough about the memory layout and reference counting of these objects.  You'll need to show me how the transmute should look.

I added the RendererExtManual trait and included it in the prelude.

I also quit using the `get_type` section of the wrapper macro, since it fulfills some other traits that I don't want.  This is just like GDK Event.

bvinc commented:
I addressed all of the feedback.

I added the RendererExtManual trait and included it in the prelude.

I also quit using the get_type section of the wrapper macro, since it fulfills some other traits that I don't want. This is just like GDK Event.

I think the only issue left to nail down is maybe renaming `IsARenderNode`?

bilelmoussaoui commented on gsk4/src/render_node.rs:
```diff
@@ -0,0 +1,173 @@
+        unsafe { from_glib_full(ffi::gsk_render_node_serialize(self.to_glib_none().0)) }
+    }
+
+    pub fn write_to_file(&self, filename: &Path) -> Result<(), glib::Error> {
```
```suggestion
    pub fn write_to_file<P: AsRef<std::path::Path>(&self, filename: &P) -> Result<(), glib::Error> {
```

bilelmoussaoui commented on gsk4/src/render_node.rs:
```diff
@@ -0,0 +1,173 @@
+            let mut error = ptr::null_mut();
+            let _ = ffi::gsk_render_node_write_to_file(
+                self.to_glib_none().0,
+                filename.to_glib_none().0,
```
```suggestion
                filename.as_ref().to_glib_none().0,
```

bvinc replied:
Thanks.  I got it working.

sdroege commented on gsk4/src/render_node.rs:
```diff
@@ -0,0 +1,173 @@
+    }
+}
+
+#[macro_export]
```
We probably don't want to export this outside the crate

bvinc replied:
There might be something that I don't understand here, but I couldn't figure out how to make a crate-scoped macro.  It seems like I can either define a macro that's valid for the rest of the file, or I can export it.  Since I really wanted to keep these render nodes in their own separate files, I was forced to export it.

bvinc replied:
https://danielkeep.github.io/tlborm/book/mbe-min-scoping.html  Ah, yes, I needed to do that, and also use `#[macro_use]` to allow the macro to escape the module.

sdroege commented on gsk4/src/render_node.rs:
```diff
@@ -0,0 +1,173 @@
+            const NODE_TYPE: RenderNodeType = $node_type;
+
+            fn upcast(self) -> crate::RenderNode {
+                unsafe { ::std::mem::transmute(self) }
```
If you prefer you can also transmute in `downcast`, or do the `from_glib_full(to_glib_full)` also here. Let's just keep it consistent in both places.

bvinc replied:
I don't know how to use transmute on downcast, since it returns a T.  I just changed it back to `from_glib_full(to_glib_full)` so it can be just like `ParamSpec`.

sdroege replied:
Ok, all the same really as this is not a performance-critical path :)

bvinc commented on gsk4/src/border_node.rs:
```diff
@@ -1,19 +1,35 @@
+
+impl BorderNode {
+    //#[doc(alias = "gsk_border_node_new")]
+    //pub fn new(outline: &RoundedRect, border_width: /*Unimplemented*/FixedArray TypeId { ns_id: 0, id: 20 }; 4, border_color: /*Unimplemented*/FixedArray TypeId { ns_id: 11, id: 79 }; 4) -> BorderNode {
```
Fixed.

sdroege commented on gsk4/src/border_node.rs:
```diff
@@ -1,19 +1,46 @@
+            from_glib_full(ffi::gsk_border_node_new(
+                outline.to_glib_none().0,
+                border_width,
+                mem::transmute(border_color),
```
```suggestion
                border_color.as_ptr() as *const gdk::ffi::GdkRGBA,
```

sdroege commented on gsk4/src/border_node.rs:
```diff
@@ -1,19 +1,46 @@
+        unsafe {
+            from_glib_full(ffi::gsk_border_node_new(
+                outline.to_glib_none().0,
+                border_width,
```
```suggestion
                border_width.as_ptr(),
```

bvinc commented:
I renamed `IsARenderNode` to `IsRenderNode` and also managed to remove the macro export.
